packages/element-web-module-api/src/api/custom-components.ts

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2025 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+Please see LICENSE files in the repository root for full details.
+*/
+
+import type { JSX } from "react";
+import type { MatrixEvent } from "matrix-js-sdk";
+
+/**
+ * Targets in Element for custom components.
+ * @public
+ */
+export enum CustomComponentTarget {
+    /**
+     * Component that renders "m.room.message" events in the room timeline.
+     */
+    TextualBody = "TextualBody",
+}
+
+/**
+ * Properties for the render component.
+ * @public
+ */
+export type CustomComponentProps = {
+    [CustomComponentTarget.TextualBody]: {
+        /**
+         * The Matrix event for this textual body.
+         */
+        mxEvent: MatrixEvent;
+        /**
+         * Words to highlight on (e.g. from search results).
+         * May be undefined if the client does not need to highlight
+         */
+        highlights?: string[];
+        /**
+         * Should previews be shown for this event
+         */
+        showUrlPreview?: boolean;
+        /**
+         * Is this event being rendered to a static export
+         */
+        forExport?: boolean;
+    };
+};
+/**
+ * Render function. Returning null skips this function and passes it onto the next registered renderer.
+ * @public
+ */
+export type CustomComponentRenderFunction<T extends CustomComponentTarget> = (
+    /**
+     * Properties from the given target to be used for rendering.
+     */
+    props: CustomComponentProps[T],
+    /**
+     * Render function for the original component.
+     */
+    originalComponent: () => JSX.Element,
+) => JSX.Element | null;
+
+/**
+ * API for inserting custom components into Element.
+ * @public
+ */
+export interface CustomComponentsApi {
+    /**
+     * Register a renderer for a component type.
+     * The render function should either return a rendered component, or `null` if the
+     * component should not be overidden.
+     *
+     * Multiple render function may be registered for a single target, however the first
+     * non-null result will be used. If all results are null, or no registrations exist
+     * for a target then the original component is used.
+     *
+     * @param target - The target location for the component.
+     * @param renderer - The render method.
+     */
+    register<T extends CustomComponentTarget>(target: T, renderer: CustomComponentRenderFunction<T>): void;
+}
```

Comment:
This has very generic naming but the Props type is very specific to timeline tiles, like if this was trying to override a button component or a settings dialog then mxEvent, highlights etc would be unrelated to the component at hand, this seems quite confused

Reply:
Oh, that's not how I wanted this to work at all. The intention is that `target` defines the prop types thought types. E.g. `TextualBody` maps to https://github.com/element-hq/element-modules/pull/26/files#diff-b159e859ecdc58575e73f5a84c28783d630caac7193b3a7912c899c7b3b44e3bR26. 

The intention being if you add more target types, you define those props too. This seemed better than adding many different register functions `registerTextualBody`, `registerButton` but if you have strong concerns then this can change.

Reply:
I think a pattern which doesn't enable the ability to extend the product, only to modify existing parts isn't flexible enough. Not being able to render events which Element Web has no renderer for